docs-chatbot/generate-embeddings.ts

```diff
@@ -0,0 +1,123 @@
+import dotenv from "dotenv";
+import { Configuration, OpenAIApi } from "openai";
+import { promises as fs } from "fs";
+import { join } from "path";
+import { encode } from "gpt-tokenizer";
+import * as edgedb from "edgedb";
+import e from "dbschema/edgeql-js";
+
+dotenv.config({ path: ".env.local" });
+
+interface Section {
+  id?: string;
+  path: string;
+  tokens: number;
+  content: string;
+  embedding: number[];
+}
+
+async function walk(dir: string): Promise<string[]> {
+  const immediateFiles = await fs.readdir(dir);
+
+  const recursiveFiles: string[][] = await Promise.all(
+    immediateFiles.map(async (file: any) => {
+      const path = join(dir, file);
+      const stats = await fs.stat(path);
+      if (stats.isDirectory()) return walk(path);
+      else if (stats.isFile()) return [path];
+      else return [];
+    })
+  );
+
+  const flattenedFiles: string[] = recursiveFiles.reduce(
+    (all, folderContents) => all.concat(folderContents),
+    []
+  );
+
+  return flattenedFiles.sort((a, b) => a.localeCompare(b));
+}
+
+async function prepareSectionsData(
+  sectionPaths: string[],
+  openai: OpenAIApi
+): Promise<Section[]> {
+  const contents: string[] = [];
+  const sections: Section[] = [];
+
+  for (const path of sectionPaths) {
+    const content = await fs.readFile(path, "utf8");
+    // OpenAI recommends replacing newlines with spaces for best results (specific to embeddings)
+    const contentTrimmed = content.replace(/\n/g, " ");
+    contents.push(contentTrimmed);
+    sections.push({
+      path,
+      content,
+      tokens: 0,
+      embedding: [],
+    });
+  }
+
+  const embeddingResponse = await openai.createEmbedding({
+    model: "text-embedding-ada-002",
+    input: contents,
+  });
+
+  if (embeddingResponse.status !== 200) {
+    throw new Error(embeddingResponse.statusText);
+  }
+
+  embeddingResponse.data.data.forEach((item, i) => {
+    sections[i].embedding = item.embedding;
+    sections[i].tokens = encode(contents[i]).length;
+  });
+
+  return sections;
+}
+
+async function storeEmbeddings() {
+  if (!process.env.OPENAI_API_KEY) {
+    return console.log(
+      "Environment variable OPENAI_API_KEY is required: skipping embeddings generation."
+    );
+  }
+
+  try {
```

Comment:
I just got to this point in editing. I was going to suggest breaking up this try…catch block, but actually, looking at what it does, I'm not sure we need it. The only thing it does best I can tell is log out the error message and allow for the script to continue, which in this case I think would just mean that "Embedding generation complete" will also get logged.

Maybe the simpler way is to just let any errors throw. This way, the user doesn't get "Embedding generation complete" if it fails (which could be misleading), and we get rid of a control flow structure that makes the code a bit harder to follow (and harder to break up in the tutorial).

Reply:
I think we should keep try/catch block. We can move the successful generation message inside the try block.

```
    await query.run(client, { sections });
    console.log("Embedding generation complete");
  } catch (err) {
    console.error("Error while trying to regenerate all embeddings.", err);
  }
}
```

Reply:
Can you explain what the try/catch does for us here? I'm having trouble understanding what it gives us which is why I suggest we remove it.

Reply:
I think that we should handle errors and not crash and terminate the program due to the errors. Usually when using try/catch block we can show more user friendly message to the user (and other things like error reporting, cleanup..., but we are not doing those here). It's a good practice to have try/catch blocks even if we can remove it here since this code runs on the BE only.

Reply:
We're not really "handling" the errors in any meaningful way though. We're just putting a generic message in front of the error and logging it all out. If there's a way to handle them, let's handle them (although maybe not in this example app since it makes things more complicated… unless it's necessary for the app to function), but I don't know that logging out a message does anything more to handle them than crashing does.

Crashing is OK here because we don't know if it is not safe to keep running the script after the error state is hit. We already caught one bug this introduced: we were showing a success message after the script had failed. You've proposed to move that now, but it doesn't really address the problem: we're "handling" an error by allowing the script to continue when it shouldn't. It would be easy for someone who isn't doesn't realize this script will catch a critical exception and continue running to introduce a new bug similar to the success log one. By *not* crashing here, the script could continue and do things we don't want it to do or that are useless. If we let it throw when it does, in fact, encounter an error that can't be meaningfully handled, we avoid that problem. We could sorta address this by catching and then re-throwing the error, but it feels pointless since all we gain is adding that sentence in front of the error. If this were an HTTP server and we needed to send a response back to the client, it would make sense to catch an error so that we could do that, but I don't see what it does for us in this context.

Then finally, we still have code that is running outside a try/catch, so if the principle we're trying to follow is that we should make sure we catch every error, we aren't doing that either. It seems to me that it just adds more complexity that the reader has to try to come to grips with, and I can't really help them by writing an explanation since I don't understand what it does for us.